File: src/envelope.py
"""ztr-finding-1 envelope validation and signing helpers (Week 1 Day 5)."""

from __future__ import annotations

import re
from datetime import datetime, timedelta, timezone
from typing import Any, Mapping, MutableMapping

from canonicalize import (
    body_digest_hex,
    canonicalize_envelope_for_signing,
    hmac_sha256_hex,
    payload_digest_hex,
    verify_hmac_sha256_hex,
)
from errors import IngestError, IngestErrorCode

VERSION = "ztr-finding-1"
ALG = "hmac-sha256"
DEFAULT_SKEW_SECONDS = 300
ISSUED_AT_Z = re.compile(
    r"^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\.[0-9]+)?Z$"
)

REQUIRED_CORE = (
    "version",
    "org_id",
    "sender_id",
    "kid",
    "alg",
    "issued_at",
    "nonce",
)


def validate_envelope_shape(
    envelope: Mapping[str, Any],
    *,
    for_signing: bool = False,
) -> None:
    """Structural checks only (no crypto)."""
    if envelope.get("version") != VERSION:
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            f"version must be {VERSION!r}",
        )

    missing = [field for field in REQUIRED_CORE if field not in envelope]
    if missing:
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            f"missing required fields: {', '.join(missing)}",
        )

    plaintext = envelope.get("payload_plaintext")
    ciphertext = envelope.get("payload_ciphertext")
    plaintext_mode = envelope.get("plaintext_mode")

    if plaintext is not None and ciphertext is not None:
        raise IngestError(
            IngestErrorCode.INVALID_PAYLOAD_MODE,
            "payload_plaintext and payload_ciphertext are mutually exclusive",
        )
    if plaintext is None and ciphertext is None:
        raise IngestError(
            IngestErrorCode.INVALID_PAYLOAD_MODE,
            "exactly one payload field is required",
        )
    if plaintext is not None:
        if plaintext_mode is not True:
            raise IngestError(
                IngestErrorCode.INVALID_PAYLOAD_MODE,
                "plaintext_mode must be true when payload_plaintext is set",
            )
        if not isinstance(plaintext, Mapping):
            raise IngestError(
                IngestErrorCode.INVALID_ENVELOPE,
                "payload_plaintext must be a JSON object",
            )
    elif plaintext_mode is True:
        raise IngestError(
            IngestErrorCode.INVALID_PAYLOAD_MODE,
            "plaintext_mode true requires payload_plaintext",
        )
    if ciphertext is not None and not isinstance(ciphertext, str):
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            "payload_ciphertext must be a base64 string",
        )

    if not for_signing:
        for field in ("payload_digest", "signature"):
            if field not in envelope:
                raise IngestError(
                    IngestErrorCode.INVALID_ENVELOPE,
                    f"missing required fields: {field}",
                )

    if envelope.get("alg") != ALG:
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            f"alg must be {ALG!r}",
        )

    issued_at = envelope.get("issued_at")
    if not isinstance(issued_at, str) or not ISSUED_AT_Z.match(issued_at):
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            "issued_at must be RFC 3339 UTC with Z suffix",
        )


def parse_issued_at_utc(value: str) -> datetime:
    if not ISSUED_AT_Z.match(value):
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            "issued_at must be RFC 3339 UTC with Z suffix",
        )
    return datetime.fromisoformat(value.replace("Z", "+00:00")).astimezone(timezone.utc)


def check_clock_skew(
    issued_at: datetime,
    *,
    now: datetime | None = None,
    window_seconds: int = DEFAULT_SKEW_SECONDS,
) -> None:
    now = now or datetime.now(timezone.utc)
    delta = abs((issued_at - now).total_seconds())
    if delta > window_seconds:
        raise IngestError(
            IngestErrorCode.CLOCK_SKEW,
            f"issued_at outside ±{window_seconds} second window",
        )


def verify_payload_digest_field(envelope: Mapping[str, Any]) -> None:
    declared = envelope.get("payload_digest")
    if not isinstance(declared, str):
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            "payload_digest must be a hex string",
        )
    plaintext = envelope.get("payload_plaintext")
    ciphertext = envelope.get("payload_ciphertext")
    try:
        expected = payload_digest_hex(
            plaintext=plaintext if isinstance(plaintext, Mapping) else None,
            ciphertext_b64=ciphertext if isinstance(ciphertext, str) else None,
        )
    except ValueError as exc:
        raise IngestError(
            IngestErrorCode.INVALID_PAYLOAD_MODE,
            str(exc),
        ) from exc
    if declared.lower() != expected:
        raise IngestError(
            IngestErrorCode.DIGEST_MISMATCH,
            "payload_digest does not match payload bytes",
        )


def parse_body_digest_header(header_value: str) -> str:
    if not header_value.startswith("sha256="):
        raise IngestError(
            IngestErrorCode.DIGEST_MISMATCH,
            "X-BLT-Body-Digest must start with sha256=",
        )
    hex_part = header_value[7:]
    if len(hex_part) != 64 or not all(c in "0123456789abcdef" for c in hex_part.lower()):
        raise IngestError(
            IngestErrorCode.DIGEST_MISMATCH,
            "X-BLT-Body-Digest must be sha256=<64 hex chars>",
        )
    return hex_part.lower()


def verify_body_digest(raw_body: bytes, header_value: str) -> None:
    declared = parse_body_digest_header(header_value)
    expected = body_digest_hex(raw_body)
    if declared != expected:
        raise IngestError(
            IngestErrorCode.DIGEST_MISMATCH,
            "request body digest does not match X-BLT-Body-Digest",
        )


def sign_envelope(envelope: MutableMapping[str, Any], secret: bytes) -> str:
    """Compute HMAC signature; does not mutate envelope except returning signature hex."""
    unsigned = {k: v for k, v in envelope.items() if k != "signature"}
    message = canonicalize_envelope_for_signing(unsigned)
    return hmac_sha256_hex(secret, message)


def verify_envelope_signature(envelope: Mapping[str, Any], secret: bytes) -> None:
    signature = envelope.get("signature")
    if not isinstance(signature, str):
        raise IngestError(
            IngestErrorCode.INVALID_ENVELOPE,
            "signature must be a hex string",
        )
    unsigned = {k: v for k, v in envelope.items() if k != "signature"}
    message = canonicalize_envelope_for_signing(unsigned)
    if not verify_hmac_sha256_hex(secret, message, signature):
        raise IngestError(
            IngestErrorCode.BAD_SIGNATURE,
            "HMAC verification failed",
        )


def prepare_signed_envelope(
    envelope: MutableMapping[str, Any],
    secret: bytes,
    *,
    now: datetime | None = None,
) -> MutableMapping[str, Any]:
    """Validate shape, set payload_digest if absent, sign, and return envelope."""
    validate_envelope_shape(envelope, for_signing=True)
    if "payload_digest" not in envelope or not envelope["payload_digest"]:
        plaintext = envelope.get("payload_plaintext")
        ciphertext = envelope.get("payload_ciphertext")
        envelope["payload_digest"] = payload_digest_hex(
            plaintext=plaintext if isinstance(plaintext, Mapping) else None,
            ciphertext_b64=ciphertext if isinstance(ciphertext, str) else None,
        )
    verify_payload_digest_field(envelope)
    issued = parse_issued_at_utc(envelope["issued_at"])
    check_clock_skew(issued, now=now)
    envelope["signature"] = sign_envelope(envelope, secret)
    return envelope

File: client/lib/ingest/envelope.dart
/// ztr-finding-1 envelope builder + HMAC signing (parity with `src/envelope.py`).
library;

import 'dart:typed_data';

import 'canonicalize.dart';

const String ztrVersion = 'ztr-finding-1';
const String ztrAlg = 'hmac-sha256';

class EnvelopeValidationException implements Exception {
  EnvelopeValidationException(this.message);
  final String message;
  @override
  String toString() => 'EnvelopeValidationException: $message';
}

/// Build a signed plaintext-mode finding envelope ready for POST /api/ingest.
Map<String, Object?> prepareSignedPlaintextEnvelope({
  required String orgId,
  required String senderId,
  required String kid,
  required Uint8List secret,
  required Map<String, Object?> payload,
  required String issuedAt,
  required String nonce,
}) {
  final envelope = <String, Object?>{
    'version': ztrVersion,
    'org_id': orgId,
    'sender_id': senderId,
    'kid': kid,
    'alg': ztrAlg,
    'issued_at': issuedAt,
    'nonce': nonce,
    'plaintext_mode': true,
    'payload_plaintext': payload,
  };
  return prepareSignedEnvelope(envelope, secret);
}

Map<String, Object?> prepareSignedEnvelope(
  Map<String, Object?> envelope,
  Uint8List secret,
) {
  _validateShape(envelope, forSigning: true);
  if (envelope['payload_digest'] == null ||
      (envelope['payload_digest'] as String).isEmpty) {
    final plaintext = envelope['payload_plaintext'];
    final ciphertext = envelope['payload_ciphertext'];
    envelope['payload_digest'] = payloadDigestHex(
      plaintext: plaintext is Map<String, Object?>
          ? Map<String, Object?>.from(plaintext)
          : (plaintext is Map
              ? plaintext.map((k, v) => MapEntry(k.toString(), v as Object?))
              : null),
      ciphertextB64: ciphertext is String ? ciphertext : null,
    );
  }
  envelope['signature'] = _signEnvelope(envelope, secret);
  return envelope;
}

/// Compact UTF-8 JSON bytes for the HTTP body (unsorted keys — wire format).
Uint8List encodeEnvelopeBody(Map<String, Object?> envelope) {
  // Use canonical encoding for stable digests in tests; wire body must be the
  // exact bytes we hash. Prefer sorted keys so Dart/Python clients agree when
  // both use sort_keys.
  return canonicalizeJson(envelope);
}

String _signEnvelope(Map<String, Object?> envelope, Uint8List secret) {
  final message = canonicalizeEnvelopeForSigning(envelope);
  return hmacSha256Hex(secret, message);
}

void _validateShape(Map<String, Object?> envelope, {required bool forSigning}) {
  if (envelope['version'] != ztrVersion) {
    throw EnvelopeValidationException('version must be $ztrVersion');
  }
  const requiredCore = [
    'version',
    'org_id',
    'sender_id',
    'kid',
    'alg',
    'issued_at',
    'nonce',
  ];
  final missing = requiredCore.where((f) => !envelope.containsKey(f)).toList();
  if (missing.isNotEmpty) {
    throw EnvelopeValidationException(
      'missing required fields: ${missing.join(', ')}',
    );
  }
  final plaintext = envelope['payload_plaintext'];
  final ciphertext = envelope['payload_ciphertext'];
  final plaintextMode = envelope['plaintext_mode'];
  if (plaintext != null && ciphertext != null) {
    throw EnvelopeValidationException(
      'payload_plaintext and payload_ciphertext are mutually exclusive',
    );
  }
  if (plaintext == null && ciphertext == null) {
    throw EnvelopeValidationException('exactly one payload field is required');
  }
  if (plaintext != null) {
    if (plaintextMode != true) {
      throw EnvelopeValidationException(
        'plaintext_mode must be true when payload_plaintext is set',
      );
    }
    if (plaintext is! Map) {
      throw EnvelopeValidationException('payload_plaintext must be a JSON object');
    }
  } else if (plaintextMode == true) {
    throw EnvelopeValidationException(
      'plaintext_mode true requires payload_plaintext',
    );
  }
  if (envelope['alg'] != ztrAlg) {
    throw EnvelopeValidationException('alg must be $ztrAlg');
  }
  if (!forSigning) {
    for (final field in ['payload_digest', 'signature']) {
      if (!envelope.containsKey(field)) {
        throw EnvelopeValidationException('missing required fields: $field');
      }
    }
  }
}

/// Hex-decode a sender secret (e.g. demo `736563726574`).
Uint8List secretFromHex(String hex) {
  final cleaned = hex.trim().replaceAll(RegExp(r'\s+'), '');
  if (cleaned.length.isOdd) {
    throw FormatException('secret hex must have even length');
  }
  final out = Uint8List(cleaned.length ~/ 2);
  for (var i = 0; i < out.length; i++) {
    out[i] = int.parse(cleaned.substring(i * 2, i * 2 + 2), radix: 16);
  }
  return out;
}

String issuedAtNowUtc() {
  final now = DateTime.now().toUtc();
  String two(int n) => n.toString().padLeft(2, '0');
  return '${now.year.toString().padLeft(4, '0')}-'
      '${two(now.month)}-${two(now.day)}T'
      '${two(now.hour)}:${two(now.minute)}:${two(now.second)}Z';
}
